cppcache/include/geode/DataOutput.hpp

```diff
@@ -495,12 +494,16 @@ class APACHE_GEODE_EXPORT DataOutput {
   /** Destruct a DataOutput, including releasing the created buffer. */
   ~DataOutput() {
     reset();
-    DataOutput::checkinBuffer(m_bytes, m_size);
+    if (m_bytes != nullptr) {
```

Comment:
Making this a unique_ptr is the fix for the double delete issue when a DataOutput instance goes away and tries to give back a duplicate buffer to the pool.  The alternatives are either a custom move ctor in DataOutput or explicit code in the TSSDataOutput class to check for duplicate entries (and null entries) before adding to the list.  This seemed like the cleanest option of the three.

Reply:
@austxcodemonkey that can't be correct. The original code never deletes the buffer, it checks it out of and into the buffer pool.

Reply:
The original code didn't use move or copy ctors for DataOutput, cause we were always just passing around the pointer.  In that case, you can never have more than one DataOutput object with the same value for m_bytes, the buffer that belongs to the pool.  With move semantics, though, and a raw pointer for m_bytes, you get duplicates.  When these DataOutput objects go away, they all give their buffer(s) back to the pool, and you get duplicate entries in the pool.  The pool dtor walks its list and deletes all the entries, so if there are duplicates you get a double delete.

Reply:
We changed this to a raw pointer, and added the necessary move ctor and move assignment operator, so let us know which you prefer.  I'm more or less agnostic, though I think I _slightly_ prefer to avoid custom move semantics.

Reply:
unique_ptr makes sense now. The thing that makes this odd is that we take ownership of a raw and ultimately release the raw. I’d suggest we convert this pool to unique_ptr so that perhaps it’s more clear what’s going on.

Reply:
[GEODE-5117](https://issues.apache.org/jira/browse/GEODE-5117) :)